### PR TITLE
Verify/unveify user

### PR DESCRIPTION
**Pull request by Rup-Narayan-Rajbanshi**
Implements https://github.com/kitab-bazar/kitab-bazar/issues/42

- Addresses User

## Changes
- Verify/Unverify User

## This PR doesn't introduce any:

- [ ] temporary files, auto-generated files or secret keys
- [ ] n+1 queries
- [ ] flake8 issues
- [ ] `print`
- [ ] typos
- [ ] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/user/serializers.py [hunk 1]
```diff
@@ -166,6 +166,28 @@ def validate(self, attrs):
         raise serializers.ValidationError(gettext('Activation link is not valid.'))
 
 
+class VerifySerializer(serializers.Serializer):
```
Comment (thenav56) [inline apps/user/serializers.py]: Let's Use ModelSerializer, and use fields = ('is_verified',)
```suggestion
class UserVerifySerializer(serializers.ModelSerializer):
```

File: apps/user/serializers.py [hunk 2]
```diff
@@ -166,6 +166,28 @@ def validate(self, attrs):
...
+        try:
+            user = User.objects.get(pk=id)
+        except (TypeError, ValueError, OverflowError, User.DoesNotExist):
+            user = None
+
+        if user is not None:
+            user.is_verified = is_verified
+            user.save()
+            return attrs
+        raise serializers.ValidationError(gettext('User with this id doesnot exists'))
```
Comment (thenav56) [inline apps/user/serializers.py]: Let's never change the database entity through validation.
We have basic mutation where you can specify custom permissions like this one. Add permission in configs/permissions.py VERIFY_USER, and then mention that permission using CreateUpdateGrapheneMutation with id. That way you don't have to retrieve the user yourself and validate user permissions.

https://github.com/kitab-bazar/server/blob/afbb95ec58e110851ca6196c8855d5aebcbb29bd/apps/order/mutations.py#L95-L102

```suggestion
    def validate(self, data):
        user_type = self.context['request'].user.user_type
        if not user_type == User.UserType.MODERATOR:
            raise serializers.ValidationError(gettext('Permission denied'))
        return data
```

File: apps/user/mutations.py [hunk 3]
```diff
@@ -107,6 +108,35 @@ def mutate(root, info, data):
         return Activate(errors=None, ok=True)
 
 
+VerifyInputType = generate_input_type_for_serializer(
+    'VerifyInputType',
+    VerifySerializer
+)
+
+
+class Verify(graphene.Mutation):
```
Comment (thenav56) [inline apps/user/mutations.py]: Using CreateUpdateMutation here
https://github.com/kitab-bazar/server/blob/afbb95ec58e110851ca6196c8855d5aebcbb29bd/apps/order/mutations.py#L95-L102

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Comment by frozenhelium** [comment]
@Rup-Narayan-Rajbanshi 
Commit message should be an Imperative Sentence

eg:
`Add model and mutation for user verification`

File: apps/user/filters.py [hunk 4]
```diff
@@ -1,15 +1,33 @@
 import django_filters
 from apps.user.models import User
+from django.db.models import Q
 
 
 class UserFilter(django_filters.FilterSet):
     full_name = django_filters.CharFilter(method='filter_full_name')
+    user_type = django_filters.CharFilter(method='filter_user_type')
```
Comment (thenav56) [inline apps/user/filters.py]: Use MultipleInputFilter instead.

https://github.com/kitab-bazar/server/blob/cc9bc6bd7af57aae21de79fdd8063c883d1a9522/apps/book/filters.py#L18

File: apps/user/filters.py [hunk 5]
```diff
@@ -1,15 +1,33 @@
...
 
     def filter_full_name(self, queryset, name, value):
         if not value:
             return queryset
         return queryset.filter(full_name__icontains=value)
+
+    def filter_user_type(self, queryset, name, value):
+        if not value:
+            return queryset
+        return queryset.filter(user_type=value)
```
Comment (thenav56) [inline apps/user/filters.py]: Not required after using MultipleInputFilter

File: apps/user/filters.py [hunk 6]
```diff
@@ -1,15 +1,33 @@
...
+
+    def filter_search(self, queryset, name, value):
+        if not value:
+            return queryset
+        return queryset.filter(
+            Q(first_name__icontains=value) |
+            Q(last_name__icontains=value) |
```
Comment (thenav56) [inline apps/user/filters.py]: Let's use this (full_name__icontains) instead of first_name and last_name

File: apps/user/filters.py [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
 
     class Meta:
         model = User
-        fields = ['email', 'is_active']
+        fields = ['email', 'is_active', 'is_verified', 'user_type']
 
     def filter_full_name(self, queryset, name, value):
         if not value:
             return queryset
         return queryset.filter(full_name__icontains=value)
```
Comment (thenav56) [inline apps/user/filters.py]: Let's remove this now.

File: apps/user/filters.py [hunk 8]
```diff
@@ -1,15 +1,33 @@
 import django_filters
 from apps.user.models import User
+from django.db.models import Q
 
 
 class UserFilter(django_filters.FilterSet):
     full_name = django_filters.CharFilter(method='filter_full_name')
```
Comment (thenav56) [inline apps/user/filters.py]: Let's remove this now.

File: apps/user/mutations.py [hunk 9]
```diff
@@ -107,6 +110,27 @@ def mutate(root, info, data):
...
+        id = graphene.ID(required=True)
+
+    model = User
+    serializer_class = UserVerifySerializer
+    result = graphene.Field(UserType)
+    permissions = [UserPermissions.Permission.CAN_VERIFY_USER]
+
+
+class UnVerify(Verify):
+    pass
```
Comment (thenav56) [inline apps/user/mutations.py]: let's remove this

File: apps/user/mutations.py [hunk 10]
```diff
@@ -107,6 +110,27 @@ def mutate(root, info, data):
         return Activate(errors=None, ok=True)
 
 
+VerifyInputType = generate_input_type_for_serializer(
+    'VerifyInputType',
+    UserVerifySerializer
+)
+
+
+class Verify(CreateUpdateGrapheneMutation):
```
Comment (thenav56) [inline apps/user/mutations.py]: class VerifyUser

File: apps/user/serializers.py [hunk 11]
```diff
@@ -166,6 +166,19 @@ def validate(self, attrs):
...
+
+    class Meta:
+        model = User
+        fields = ('id', 'is_verified')
+
+    def validate(self, attrs):
+        user_type = self.context['request'].user.user_type
+        if not user_type == User.UserType.MODERATOR:
+            raise serializers.ValidationError(gettext('Permission denied'))
+        return attrs
```
Comment (thenav56) [inline apps/user/serializers.py]: This is not required now.

**Review by thenav56 — CHANGES_REQUESTED** [review]
Let's
- Add tests.
- Add verfied_by as well

**Review by thenav56 — APPROVED** [review]

**Pull request merged**